tests/hkdf.cc

```diff
@@ -0,0 +1,239 @@
+ */
+
+#include <fstream>
+
+#include <gtest/gtest.h>
+#include <nlohmann/json.hpp>
+
+#include "Hacl_HKDF.h"
+#include "Hacl_Spec.h"
+#include "util.h"
+
+using json = nlohmann::json;
+
+typedef struct
+{
+  std::string comment;
+  // A tcId of 0 indicates "not Wycheproof".
+  uint32_t tcId;
+  std::string hash;
+  // 1) Input Keying Material
+  bytes IKM;
+  bytes salt;
+  bytes info;
+  // Length of OKM
+  uint32_t L;
+  // 2) A pseudorandom key (of HashLen octets)
+  // Only available when "not Wycheproof".
+  bytes PRK;
+  // 3) Output Keying Material
+  bytes OKM;
+  bool valid;
+} TestCase;
+
+class HKDFSuite : public ::testing::TestWithParam<TestCase>
+{};
+
+TEST_P(HKDFSuite, TestCase)
+{
+  auto test = GetParam();
+  std::cout << "# " << test.comment << std::endl;
+
+  // TODO: How does error handling in Hacl_HKDF_* work?
+  if (!test.valid) {
+    return;
+  }
+
+  if (test.hash == "SHA-1") {
+    std::cout << "Skipping... (SHA-1 not supported)" << std::endl;
+  } else if (test.hash == "SHA-256") {
+    bytes got_prk = std::vector<uint8_t>(32);
+
+    Hacl_HKDF_extract_sha2_256(got_prk.data(),
+                               test.salt.data(),
+                               test.salt.size(),
+                               test.IKM.data(),
+                               test.IKM.size());
+
+    if (test.tcId == 0) {
+      EXPECT_EQ(test.PRK, got_prk);
+    }
+
+    bytes got_OKM = std::vector<uint8_t>(test.L);
+
+    Hacl_HKDF_expand_sha2_256(got_OKM.data(),
+                              got_prk.data(),
+                              got_prk.size(),
+                              test.info.data(),
+                              test.info.size(),
+                              test.L);
+
+    EXPECT_EQ(test.OKM, got_OKM);
+  } else if (test.hash == "SHA-512") {
+    bytes got_prk = std::vector<uint8_t>(64);
+
+    Hacl_HKDF_extract_sha2_512(got_prk.data(),
+                               test.salt.data(),
+                               test.salt.size(),
+                               test.IKM.data(),
+                               test.IKM.size());
+
+    if (test.tcId == 0) {
+      EXPECT_EQ(test.PRK, got_prk);
+    }
+
+    bytes got_OKM = std::vector<uint8_t>(test.L);
+
+    Hacl_HKDF_expand_sha2_512(got_OKM.data(),
+                              got_prk.data(),
+                              got_prk.size(),
+                              test.info.data(),
+                              test.info.size(),
+                              test.L);
+
+    EXPECT_EQ(test.OKM, got_OKM);
+  } else if (test.hash == "blake2s") {
+    FAIL()
+      << "This should not run. We have no test vectors for HKDF-BLAKE2S yet.";
+  } else if (test.hash == "blake2b") {
+    FAIL()
+      << "This should not run. We have no test vectors for HKDF-BLAKE2B yet.";
+  } else {
+    FAIL() << "Unexpected value \"" << test.hash << "\" for `test.hash`";
+  }
+}
+
+std::vector<TestCase>
+read_json_rfc5869(char* path)
```

Comment:
The tests appear to be in the wycheproof json files as well. Is the added PRK the only difference?

Reply:
Yes. The PRK is basically the only reason I keep the vectors from the RFC.

Reply:
Is this conversation resolved or is there anything you like to be changed?